### PR TITLE
TcpProxy::CombinedUpstream: Implement watermark callback functions

### PR DESCRIPTION
**Pull request by vikaschoudhary16**
followup to #32991
Commit Message: Implement watermark functions which was a TODO and some cleanup
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: test/integration/filters/stop_iteration_and_continue_filter.cc [hunk 1]
```diff
@@ -64,6 +65,9 @@ class StopIterationAndContinueFilter : public Http::PassThroughFilter {
     if (end_stream) {
       setEndStreamAndDecodeTimer();
     }
+    if (stop_and_buffer_) {
+      return Http::FilterDataStatus::StopIterationAndBuffer;
```
Comment (vikaschoudhary16) [inline test/integration/filters/stop_iteration_and_continue_filter.cc:69]: If this change is not added, [assertion](https://github.com/envoyproxy/envoy/blob/main/source/common/tcp_proxy/tcp_proxy.cc#L760) fails in tcp_proxy.cc 
```
 // The upstream should consume all of the data.
  // Before there is an upstream the connection should be readDisabled. If the upstream is
  // destroyed, there should be no further reads as well.
  ASSERT(0 == data.length());
```

Why we dont need this assertion in HCM case and it is there only in tcp_proxy?

Comment (alyssawilk) [reply test/integration/filters/stop_iteration_and_continue_filter.cc:69]: StopIterationNobuffer is generally called before we quit processing a request.  it results in discarding data that's in the filter.
in the tcp proxy case it means the filter manager doesn't consume the data to

**Review by alyssawilk — COMMENTED** [review]
LGTM with a few questions - just to make sure you verified your tests hit the backup and unwind code?

File: source/common/tcp_proxy/upstream.cc [hunk 2]
```diff
@@ -441,9 +441,7 @@ bool CombinedUpstream::readDisable(bool disable) {
   if (upstream_requests_.empty()) {
     return false;
   }
-  if (disable) {
-    upstream_requests_.front()->onAboveWriteBufferHighWatermark();
-  }
+  upstream_requests_.front()->readDisableOrDefer(disable);
```
[the remark this answers is not in the document]
Comment (vikaschoudhary16) [reply source/common/tcp_proxy/upstream.cc:444]: yes, this will always be single entry

**Comment by vikaschoudhary16** [comment]
> LGTM with a few questions - just to make sure you verified your tests hit the backup and unwind code?

yes, I have verified that it hits backup and unwind code.

**Review by alyssawilk — APPROVED** [review]

**Pull request merged**